### ISSUE TITLE
Connect with custom modules

### ISSUE THREAD

**Issue by bickus** [blak3r/yaai#86]
Does anybody connect yaai with custom modules? Not only with contacts and accounts.

**Comment by blak3r** [comment]
No, it's a feature we'd like to support... well not sure about custom modules.  But, leads module would be good.

**Comment by bickus** [comment]
Maybe somebody could give me a hint about connecting my custom module to yaai?

**Comment by blak3r** [comment]
You'll have to become pretty much an expert in the codebase.
1. Find every reference to contact in asteriskLogger and write corresponding methods to do searches in your module.
2. Rewrite callListener.php to query from your module + contacts.  Then, rewrite the display logic to handle things like multiple matching contacts and multiple matching your module...

If you didn't need contact support you could probably just replace all the references to contact with your module and it wouldn't be too difficult.

**Comment by bickus** [comment]
Well, now i'm know where to digg. Will try! Thank you.

**Comment by askhogan** [comment]
Vladamir.  I would wait for the v3 beta or look at my fork of the project.
The templating was refactored into a handlebars.js template... Wouldnt be
too difficult to add in another field.   What custom module are you
thinking about adding? Are you wanting to add data to the callbox or
extract data from the call to your custom module?  If it is the latter, I
added an action dropdown last night that will make popping to custom module
URLs trivial if all you are wanting is the phone number output

**Comment by bickus** [comment]
Sorry for so long response. I have some custom modules with clients data (same to accounts and contacts). I want yaai to find calling number in my modules and show it (as it's for accounts & contacts) for each incoming call.
And in case of no matches for calling number - "create new" should be fired for one of my modules (customer can choose what exactly)

**Comment by blak3r** [comment]
Here's some more notes:

One issue with supporting more fields is it will make somewhat slow call queries even more slow...

Here's an excerpt from callListener:

```
    $selectPortion =  "SELECT c.id as contact_id, first_name,   last_name,phone_work, phone_home, phone_mobile, phone_other, a.name as account_name, account_id "
        . " FROM contacts c "
        . " left join accounts_contacts ac on (c.id=ac.contact_id) and (ac.deleted='0' OR ac.deleted is null)"
        . " left join accounts a on (ac.account_id=a.id) and (a.deleted='0' or a.deleted is null)";

    if( $row['contact_id'] ) {
        $wherePortion = " WHERE c.id='{$row['contact_id']}' and c.deleted='0'";
        //log_entry("Quick WHERE $selectPortion $wherePortion\n", "c:\callListenerLog.txt");
    }
    // We only do this expensive query if it's not already set!
    else {
        //log_entry("Expensive WHERE " . "\n", "c:\callListenerLog.txt");
        $wherePortion = " WHERE (";
        $wherePortion .= sprintf($sqlReplace, "phone_work", $phoneToFind) . " OR ";
        $wherePortion .= sprintf($sqlReplace, "phone_home", $phoneToFind) . " OR ";
        $wherePortion .= sprintf($sqlReplace, "phone_other", $phoneToFind) . " OR ";
        $wherePortion .= sprintf($sqlReplace, "assistant_phone", $phoneToFind) . " OR ";
        $wherePortion .= sprintf($sqlReplace, "phone_mobile", $phoneToFind) . ") and c.deleted='0'";
    }

    $queryContact = $selectPortion . $wherePortion;
    //log_entry($queryContacfastert . "\n","c:\callListenerLog.txt");
    $innerResultSet = $current_user->db->query($queryContact, false);
```

That query will need to be modified to Join in other tables and also add additional WHERE clauses.

With respect to Custom Modules... We'd need to implement a UI to allow people to:
a) Pick the Modules they want to match on.... and 
b) Then pick the fields inside that module... 

Then we'd need to update the queries to dynamically join those other tables and add the other where's.  tricky UI to dynamically pull in other tables as well and then need to handle special logic for the Multiple contact found case such that when you click on the radio button it passes the appropriate module name.
